### PR TITLE
AppSource Gallery follow-up fixes

### PR DESCRIPTION
**Pull request by blrobl**
**Summary**
 
AppSource Marketplace Iframe is being deprecated, so we are creating our own solution to replace it. This is a follow-up PR that fixes a few bugs.

**Work Item(s)**

Fixes [AB#467161](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/467161)

### CODE REVIEW

File: src/System Application/App/AppSource Gallery/app.json [hunk 1]
```diff
@@ -63,7 +63,7 @@
     ],   
     "internalsVisibleTo":  [
         {
-            "name": "Microsoft AppSource Product Gallery Test",
+            "name": "AppSource Product Gallery Test",
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/app.json:66]: still not pointing to a test library app

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 2]
```diff
@@ -576,26 +500,9 @@ codeunit 2515 "AppSource Product Manager" implements "IAppSource Product Manager
 
     local procedure SetDefaultDependencyImplementation()
     var
-        dependencyInstance: Codeunit "AppSource Product Manager";
+        DependencyInstance: Codeunit "AppSource Product Manager";
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]:
```suggestion
        AppSourceProductManager: Codeunit "AppSource Product Manager";
```

if you're unhappy with the names of your variables, change the name of the object...

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 3]
```diff
@@ -282,7 +371,7 @@ codeunit 2515 "AppSource Product Manager" implements "IAppSource Product Manager
         SetCommonHeaders(RestClient);
         RestClient.SetDefaultRequestHeader('client-request-id', ClientRequestID);
 
-        exit(Dependencies.RestClient_GetAsJSon(RestClient, RequestUri).AsObject());
+        exit(Dependencies.GetAsJSon(RestClient, RequestUri).AsObject());
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: So now i see Dependencies. all over the code and i have no idea what it is... Please name your variables after the thing it points to...

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 4]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
-codeunit 2515 "AppSource Product Manager" implements "IAppSource Product Manager Dependencies"
+codeunit 2515 "AppSource Product Manager" implements "AppSource Product Manager Dependencies"
 {
     Access = Internal;
     InherentEntitlements = X;
     InherentPermissions = X;
 
+    var
+        AppSourceJsonUtilities: Codeunit "AppSource Json Utilities";
+        Dependencies: Interface "AppSource Product Manager Dependencies";
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]:
```suggestion
        AppSourceProductManagerDependencies: Interface "AppSource Product Manager Dependencies";
```

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 5]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
+        IsDependenciesSet: boolean;
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]:
```suggestion
        IsDependenciesInterfaceSet: boolean;
```

maybe?

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 6]
```diff
@@ -374,180 +459,19 @@ codeunit 2515 "AppSource Product Manager" implements "IAppSource Product Manager
...
-                NotSupportedNotification.Send();
-
-                exit('en');
-            end;
-        end;
-    end;
-
-    #endregion
-
     local procedure InsertProductFromObject(Offer: JsonObject; var Product: Record "AppSource Product")
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]:
```suggestion
    local procedure InsertProductFromObject(Offer: JsonObject; var AppSourceProduct: Record "AppSource Product")
```

File: src/System Application/App/AppSource Gallery/app.json (hunk 1 above)
Comment (SBalslev) [reply src/System Application/App/AppSource Gallery/app.json:66]: correct, focus is bug fixing in this iteration.

File: src/System Application/Test/AppSource Gallery/src/AppSourceProductManagerTest.Codeunit.al [hunk 7]
```diff
@@ -10,7 +10,7 @@ using System.RestClient;
 using System.Environment.Configuration;
 using System.TestLibraries.Utilities;
 
-codeunit 135074 "AppSource Product Manager Test" implements "IAppSource Product Manager Dependencies"
+codeunit 135074 "AppSource Product Manager Test" implements "AppSource Product Manager Dependencies"
```
Comment (grobyns) [inline src/System Application/Test/AppSource Gallery/src/AppSourceProductManagerTest.Codeunit.al:13]: still need to separate the test code from the implementation code

File: src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al [hunk 8]
```diff
@@ -226,13 +227,14 @@ page 2516 "AppSource Product Details"
...
+        EmptyGuid: Guid;
     begin
         ProductObject := ToProductObject;
-        UniqueProductID := AppSourceProductManager.GetStringValue(ProductObject, 'uniqueProductId');
+        UniqueProductID := AppSourceJsonUtilities.GetStringValue(ProductObject, 'uniqueProductId');
         AppId := AppSourceProductManager.ExtractAppIDFromUniqueProductID(UniqueProductID);
         CurrentRecordCanBeUninstalled := false;
         CurrentRecordCanBeInstalled := false;
-        if (AppID <> '') then
+        if (AppID <> EmptyGuid) then
```
Comment (darjoo) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductDetails.Page.al]: Use `not IsNullGuid(AppID)` instead.

File: src/System Application/App/AppSource Gallery/src/AppSourceProductList.Page.al [hunk 9]
```diff
@@ -192,17 +183,28 @@ page 2515 "AppSource Product List"
...
+        Rec.SetCurrentKey(DisplayName);
     end;
 
     trigger OnAfterGetCurrRecord()
+    var
+        EmptyGuid: Guid;
     begin
         CurrentRecordCanBeUninstalled := false;
-        if (Rec.AppID <> '') then
+        if (Rec.AppID <> EmptyGuid) then
```
Comment (darjoo) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductList.Page.al]: Same here, use `not IsNullGuid(Rec.AppID)`

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 10]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
+        CatalogApiSelectQueryParamNameLbl: Label '$select', Locked = true;
+        AppSourceListingUriLbl: Label 'https://appsource.microsoft.com/%1/product/dynamics-365-business-central/%2', Comment = '%1=Language ID, such as en-US, %2=Url Query Content', Locked = true;
+        AppSourceUriLbl: Label 'https://appsource.microsoft.com/%1/marketplace/apps?product=dynamics-365-business-central', Comment = '1%=Language ID, such as en-US', Locked = true;
+        NotSupportedOnPremisesErrorLbl: Label 'Not supported on premises.';
+        UnsupportedLanguageNotificationLbl: Label 'Language %1 is not supported by AppSource. Defaulting to "en". Change the language in the user profile to use another language.', Comment = '%1=Language ID, such as en';
+        UnsupportedMarketNotificationLbl: Label 'Market %1 is not supported by AppSource. Defaulting to "us". Change the region in the user profile to use another market.', Comment = '%1=Market ID, such as "us"';
+
     #region Dependency Interface implementation
-    procedure AzureADTenant_GetCountryLetterCode(): Text[2]
+    procedure GetCountryLetterCode(): Text[2]
```
Comment (darjoo) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Code[2]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 11]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
     var
         entraTenant: Codeunit "Azure AD Tenant";
```
Comment (darjoo) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Uppercase first letter

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 12]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
     begin
         exit(entraTenant.GetCountryLetterCode());
     end;
 
-    procedure AzureAdTenant_GetPreferredLanguage(): Text
+    procedure GetPreferredLanguage(): Text
     var
         entraTenant: Codeunit "Azure AD Tenant";
```
Comment (darjoo) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Uppercase first letter

**Review by SBalslev — APPROVED** [review]

**Review by darjoo — APPROVED** [review]

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 13]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
+codeunit 2515 "AppSource Product Manager" implements "AppSource Product Manager Dependencies"
 {
     Access = Internal;
     InherentEntitlements = X;
     InherentPermissions = X;
 
+    var
+        AppSourceJsonUtilities: Codeunit "AppSource Json Utilities";
+        AppSourceProductManagerDependencies: Interface "AppSource Product Manager Dependencies";
+        IsDependenciesIntrfaceSet: boolean;
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]:
```suggestion
        IsDependenciesInterfaceSet: boolean;
```

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 14]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
     var
-        entraTenant: Codeunit "Azure AD Tenant";
+        EntraTenant: Codeunit "Azure AD Tenant";
     begin
-        exit(entraTenant.GetAadTenantID());
+        exit(EntraTenant.GetAadTenantID());
     end;
 
-    procedure AzureKeyVault_GetAzureKeyVaultSecret(SecretName: Text; var Secret: SecretText);
+    procedure GetAzureKeyVaultSecret(SecretName: Text; var Secret: SecretText);
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: This seems like a bad idea... 'Internal' is not secure and this method allows retrieval of any secret... Why do you need this helper method? 
and if you need it, it should be marked as [Scope('OnPrem')]

Comment (grobyns) [reply src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: @WaelAbuSeada wanna take a look?

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 15]
```diff
@@ -15,100 +15,117 @@ using System.RestClient;
...
+    procedure GetUserSettings(UserSecurityId: Guid; var TempUserSettingsRecord: record "User Settings" temporary)
     var
         UserSettings: Codeunit "User Settings";
     begin
         UserSettings.GetUserSettings(Database.UserSecurityID(), TempUserSettingsRecord);
     end;
 
     #endregion
-    procedure SetDependencies(SpecificDependencies: Interface "IAppSource Product Manager Dependencies")
+    procedure SetDependencies(SpecificDependencies: Interface "AppSource Product Manager Dependencies")
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al:96]: this still seems like something you only use for testing... and its a bad practice to have these backdoors for testing only.

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al [hunk 16]
```diff
@@ -150,96 +168,75 @@ codeunit 2515 "AppSource Product Manager" implements "IAppSource Product Manager
...
+    end;
 
-        RestClient.Initialize();
-        SetCommonHeaders(RestClient);
+    procedure ResolveMarketAndLanguage(var Market: Code[2]; var LanguageName: Text)
+    var
+        Language: Codeunit Language;
+        LanguageID, LocalID : integer;
+    begin
+        GetCurrentUserLanguageAnLocaleID(LanguageID, LocalID);
```
Comment (grobyns) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: there's a typo in the function name

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al (hunk 14 above)
Comment (SBalslev) [reply src/System Application/App/AppSource Gallery/src/AppSourceProductManager.Codeunit.al]: Good point about onprem scope. The current dependency management requires an implementation of the keyvault dependency.

File: src/System Application/App/AppSource Gallery/src/AppSourceProductManagerDependencies.Interface.al [hunk 17]
```diff
@@ -11,28 +11,28 @@ using System.RestClient;
...
-    procedure AzureADTenant_GetAadTenantID(): Text
-    procedure AzureADTenant_GetCountryLetterCode(): Text[2];
-    procedure AzureAdTenant_GetPreferredLanguage(): Text
+    procedure GetAadTenantID(): Text
+    procedure GetCountryLetterCode(): Code[2];
+    procedure GetPreferredLanguage(): Text
 
     // Dependency to  Azure Key Vault 
-    procedure AzureKeyVault_GetAzureKeyVaultSecret(SecretName: Text; var Secret: SecretText);
+    procedure GetAzureKeyVaultSecret(SecretName: Text; var Secret: SecretText);
```
Comment (WaelAbuSeada) [inline src/System Application/App/AppSource Gallery/src/AppSourceProductManagerDependencies.Interface.al]: This needs to be changed to directly get the needed secret and store in a "secret text"

**Review by WaelAbuSeada — APPROVED** [review]

**Review by SBalslev — APPROVED** [review]

**Review by grobyns — APPROVED** [review]
still waiting on the PR that adds the test library app

**Pull request merged**